### PR TITLE
#1448 benjamin zhu add ability to edit vendors

### LINKED ISSUE #1448: Finance Fixes - [Backend] Add ability to edit vendors
Epic: #1274 

### Description

Currently, a user is able to edit existing account codes on the admin tools page, but cannot edit vendors. This ticket is to add an edit vendor endpoint to the backend.

### Acceptance Criteria

- [ ] This endpoint should take just a name: string, as its payload
- [ ] Only admins, or finance leads, heads, or admins should be able to edit vendors
- [ ] The endpoint should throw a 400 HTTPException if someone tries to change the vendor name to a name that already exists in the database
- [ ] Unit tests for this endpoint should be added to the backend

### Proposed Solution

Look at the create vendor and edit account code endpoints for examples.

### PR DESCRIPTION
**Pull request by Bennyzhu1**
## Changes

Added a "edit vendors" endpoint that allows a user to edit the name of their vendor

## Notes
Postman Working (old vendor was named Tesla):
image

## Test Cases

- User does not have sufficient privileges to edit
- the name that User wants to change already exists
- Name successfully changed

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://nerdocs.atlassian.net/wiki/spaces/NER/pages/8060929/Software+Contributor+Guide) and reach out to your Tech Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [x] All commits are tagged with the ticket number
- [x] No linting errors / newline at end of file warnings
- [x] All code follows repository-configured prettier formatting
- [x] No merge conflicts
- [x] All checks passing
- [x] Screenshots of UI changes (see Screenshots section)
- [x] Remove any non-applicable sections of this template
- [x] Assign the PR to yourself
- [x] No `yarn.lock` changes (unless dependencies have changed)
- [x] Request reviewers & ping on Slack
- [x] PR is linked to the ticket (fill in the closes line below)

Closes #1448

### CODE REVIEW

File: src/backend/src/routes/reimbursement-requests.routes.ts [hunk 1]
```diff
@@ -24,6 +24,8 @@ reimbursementRequestsRouter.get('/reimbursements/current-user', ReimbursementReq
 
 reimbursementRequestsRouter.get('/reimbursements', ReimbursementRequestController.getAllReimbursements);
 
+reimbursementRequestsRouter.post('/edit-vendors', nonEmptyString(body('name')), ReimbursementRequestController.editVendor);
```
Comment (callmejim1226) [inline src/backend/src/routes/reimbursement-requests.routes.ts:27]: We want to add the `validateInputs` function here when there's at least one input from this endpoint payload. The `validateInputs` function makes sure there is no error from the endpoint request.

Comment (callmejim1226) [inline src/backend/src/routes/reimbursement-requests.routes.ts:27]: Also, let's update the endpoint path to `/:vendorId/vendors/edit` so we can get vendorId from `req.params` in the edit vendor controller function.

File: src/backend/src/services/reimbursement-requests.services.ts [hunk 2]
```diff
@@ -692,4 +692,31 @@ export default class ReimbursementRequestService {
...
+   * @returns new vendor with a edited name
+   */
+  static async editVendors(name: string, vendorId: string, submitter: User) {
+    if (!isAdmin(submitter.role)) throw new AccessDeniedAdminOnlyException('only Admins can edit vendors');
+
+    const vendorExists = await prisma.vendor.findUnique({
+      where: { name }
+    });
+
+    console.log(vendorExists);
```
Comment (callmejim1226) [inline src/backend/src/services/reimbursement-requests.services.ts:711]: Whoops, don't forget to remove all consoles.log statement after debugging :)

File: src/backend/src/services/reimbursement-requests.services.ts [hunk 3]
```diff
@@ -692,4 +692,31 @@ export default class ReimbursementRequestService {
...
+
+    if (!!vendorExists) throw new HttpException(400, 'vendor name already exists');
+
+    const vendor = await prisma.vendor.update({
+      where: { vendorId },
+      data: { name }
+    });
+
+    return vendor;
```
Comment (callmejim1226) [inline src/backend/src/services/reimbursement-requests.services.ts:720]: When we return the vendor object to the front end, we want to transform the type of it from `Prisma` to `Shared` type. To do this, we can wrap this vendor variable with `vendorTransformer`.

File: src/backend/tests/test-data/reimbursement-requests.test-data.ts [hunk 4]
```diff
@@ -22,6 +22,12 @@ export const PopEyes: PrismaVendor = {
   name: 'Pop Eyes'
 };
 
+export const KFC: PrismaVendor = {
+  vendorId: 'CHICKEN',
+  dateCreated: new Date('12/22/203'),
+  name: 'kfc'
```
Comment (callmejim1226) [inline src/backend/tests/test-data/reimbursement-requests.test-data.ts:28]: I wanna have KFC. It's been a while ngl lol

**Review by callmejim1226 — CHANGES_REQUESTED** [review]
Great start!! Nice endpoint. 

I left some comments on service function documentation and route definition. 
Also, let's test this endpoint manually through `Postman` and provide a screenshot of the endpoint's successful case. Lmk if you have any questions about using `Postman`.

File: src/backend/src/services/reimbursement-requests.services.ts [hunk 5]
```diff
@@ -692,4 +692,29 @@ export default class ReimbursementRequestService {
...
+  /**
+   * Edits the vendor name
+   *
+   * @param name the new vendor name
+   * @param vendorId the requested vendor to be edited
+   * @param submitter the user editing the vendor name
+   * @returns the updated vendor
+   */
+  static async editVendors(name: string, vendorId: string, submitter: User) {
+    if (!isAdmin(submitter.role)) throw new AccessDeniedAdminOnlyException('only Admins can edit vendors');
```
Comment (callmejim1226) [inline src/backend/src/services/reimbursement-requests.services.ts:705]: When I read the [ticket](https://github.com/Northeastern-Electric-Racing/FinishLine/issues/1448), it says the finance leads and heads can also edit vendors. Can we add this permission on this line?

File: src/backend/tests/reimbursement-requests.test.ts [hunk 6]
```diff
@@ -625,4 +626,28 @@ describe('Reimbursement Requests', () => {
...
+      );
+    });
+
+    test('Successfuly changes Vendors name', async () => {
+      vi.spyOn(prisma.vendor, 'update').mockResolvedValue(KFC);
+      vi.spyOn(prisma.vendor, 'findUnique').mockResolvedValue(null);
+
+      const vendor = await ReimbursementRequestService.editVendors('kfc', PopEyes.vendorId, batman);
+
+      expect(vendor.name).toBe('kfc');
```
Comment (callmejim1226) [inline src/backend/tests/reimbursement-requests.test.ts:650]: Can we also add an expected line where it checks the Prisma.vendor.update has been called 1 time? This is a way we test if our endpoint works as we expected by using Prisma ORM.

For example, we can refer to `toggles successfully` test case in projects.tests.ts

File: src/backend/tests/reimbursement-requests.test.ts [hunk 7]
```diff
@@ -625,4 +626,28 @@ describe('Reimbursement Requests', () => {
       expect(newReimbursement).toStrictEqual(reimbursementTransformer(reimbursementMock));
     });
   });
+
+  describe('Edit Vendor Tests', () => {
+    test('Throws error if user isnt an admin', async () => {
```
Comment (callmejim1226) [inline src/backend/tests/reimbursement-requests.test.ts:631]: Let's update this test name after we update the permission in the service function.

File: src/backend/tests/reimbursement-requests.test.ts [hunk 8]
```diff
@@ -625,4 +626,28 @@ describe('Reimbursement Requests', () => {
...
+      await expect(
+        ReimbursementRequestService.editVendors('I Love Benny', GiveMeMyMoney.vendorId, wonderwoman)
+      ).rejects.toThrow(new AccessDeniedAdminOnlyException('only Admins can edit vendors'));
+    });
+
+    test('Vendor Name already exists', async () => {
```
Comment (callmejim1226) [inline src/backend/tests/reimbursement-requests.test.ts:637]: Small thing, but let's change this test name to 'Throws error if the vendor name already exists', so we can have consistent test names

**Review by callmejim1226 — CHANGES_REQUESTED** [review]
Great work! Left a few comments on the service function and tests. Almost there! 👍

File: src/backend/src/services/reimbursement-requests.services.ts [hunk 9]
```diff
@@ -692,4 +692,29 @@ export default class ReimbursementRequestService {
...
+  /**
+   * Edits the vendor name
+   *
+   * @param name the new vendor name
+   * @param vendorId the requested vendor to be edited
+   * @param submitter the user editing the vendor name
+   * @returns the updated vendor
+   */
+  static async editVendors(name: string, vendorId: string, submitter: User) {
+    if (!isHead(submitter.role)) throw new AccessDeniedAdminOnlyException('only Admins and Heads can edit vendors');
```
Comment (callmejim1226) [inline src/backend/src/services/reimbursement-requests.services.ts:705]: From the ticket description, only the general admin or finance team head/leads can edit vendors. In other words, we want to check if the submitter.role is at least in an admin role, or if the submitter is in the finance team as a lead or head.

File: src/backend/tests/reimbursement-requests.test.ts [hunk 10]
```diff
@@ -625,4 +626,29 @@ describe('Reimbursement Requests', () => {
       expect(newReimbursement).toStrictEqual(reimbursementTransformer(reimbursementMock));
     });
   });
+
+  describe('Edit Vendor Tests', () => {
+    test('Throws error if user isnt an admin or head', async () => {
```
Comment (callmejim1226) [inline src/backend/tests/reimbursement-requests.test.ts:631]: Let's update the test name to say if the user isn't an admin or finance leads/head

**Review by callmejim1226 — CHANGES_REQUESTED** [review]
Almost there! 👍

File: src/backend/src/services/reimbursement-requests.services.ts [hunk 11]
```diff
@@ -692,4 +692,34 @@ export default class ReimbursementRequestService {
...
+  /**
+   * Edits the vendor name
+   *
+   * @param name the new vendor name
+   * @param vendorId the requested vendor to be edited
+   * @param submitter the user editing the vendor name
+   * @returns the updated vendor
+   */
+  static async editVendors(name: string, vendorId: string, submitter: User) {
+    if (!isAdmin(submitter.role)) {
```
Comment (callmejim1226) [inline src/backend/src/services/reimbursement-requests.services.ts:705]: Almost there! To properly check the user permission for the edit vendors, let's do the following:
1. use `await validateUserIsPartOfFinanceTeam(submitter);` to first check if the user is in finance team as either lead or head by using try and catch.
2. In a case when the user is not in finance team, then we just check if the user's role is at least admin. (in a catch case of try & catch). Note that if `!isAdmin(submitter.role)` is true, then we know `!isHead(submitter.role)` is true. You can take a look at the definition of these permission check utils functions.
3. Lastly, if the user is not at least an admin, then we gonna throw the AccessDenied exception, where the msg will be only the admin or finance lead or head can edit vendors.
4. Make this permission check as a helper method in RR utils file. Similar example can be found here for edit RR user permission check: https://github.com/Northeastern-Electric-Racing/FinishLine/pull/1672/files#diff-fc838d062bcc6c9ab4a0798a759ab857af3bb5e94e9ef0485d0881a9a19eef63

**Review by callmejim1226 — CHANGES_REQUESTED** [review]
After the user permission check fix, this PR should be good to go for the final review from Reid.

**Review by RChandler234 — APPROVED** [review]
LGTM 🔥

**Pull request merged**